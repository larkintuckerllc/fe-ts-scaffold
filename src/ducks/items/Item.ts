import { Record } from 'immutable';

const itemDefault = {
  id: 0,
  name: 'name',
};

export interface ItemJS {
  id: number;
  name: string;
}

export default class Item extends Record(itemDefault) {
  constructor(params: ItemJS) {
    super(params);
  }
  public get<T extends keyof ItemJS>(value: T): ItemJS[T] {
    return super.get(value, null);
  }
}
