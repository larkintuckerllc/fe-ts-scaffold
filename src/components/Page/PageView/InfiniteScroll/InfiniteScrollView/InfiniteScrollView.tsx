import Item from 'DUCKS/items/Item';
import { List } from 'immutable';
import React, { Component } from 'react';
import Items from './Items';

const GAP = 100;

interface InfiniteViewProps {
  currentPage: number;
  error: boolean;
  requested: boolean;
  lastPage: number;
  items: List<Item>;
  fetchItems(page: number): Promise<void>;
}

export default class InfiniteScrollView extends Component<InfiniteViewProps> {
  private rootRef: HTMLDivElement;

  public componentDidMount() {
    const { currentPage, fetchItems } = this.props;
    fetchItems(currentPage);
    window.scroll(0, 0);
    window.addEventListener('scroll', this.handleScroll, false);
  }

  // TESTING TOO HARD
  /* istanbul ignore next */
  public componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll, false);
  }

  public render() {
    const { error, items } = this.props;
    if (error) {
      return <div>Error</div>;
    }
    return (
      <div ref={this.setRootRef}>
        <h2>Infinite Scroll</h2>
        <Items items={items.toJS()} />
      </div>
    );
  }

  // TESTING TOO HARD
  /* istanbul ignore next */
  private setRootRef = (element: HTMLDivElement) => {
    this.rootRef = element;
  };

  // TESTING TOO HARD
  /* istanbul ignore next */
  private handleScroll = () => {
    const { currentPage, fetchItems, lastPage, requested } = this.props;
    const scrollY = window.scrollY;
    const windowH = window.innerHeight;
    const contentH = this.rootRef.scrollHeight;
    if (windowH + scrollY > contentH - GAP && currentPage !== lastPage && !requested) {
      fetchItems(currentPage + 1).then(this.handleScroll);
    }
  };
}
