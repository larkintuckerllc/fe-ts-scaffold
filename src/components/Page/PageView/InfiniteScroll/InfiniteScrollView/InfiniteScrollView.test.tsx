import Item from 'DUCKS/items/Item';
import Enzyme, { shallow } from 'enzyme';
import enzymeAdapterReact16 from 'enzyme-adapter-react-16';
import { List } from 'immutable';
import React from 'react';
import InfiniteScrollView from './InfiniteScrollView';

Enzyme.configure({ adapter: new enzymeAdapterReact16() });

const itemDefault = {
  id: 0,
  name: 'name',
};
const sampleItem = new Item(itemDefault);
const sampleItems = List([sampleItem]);
const getDefaultProps = () => ({
  currentPage: 0,
  error: false,
  fetchItems: jest.fn(),
  items: sampleItems,
  lastPage: 0,
  requested: false,
});

describe('InfiniteScrollView component', () => {
  window.scroll = jest.fn();
  window.addEventListener = (type: string, handler, flag: boolean) => {
    
  };
  
  ('scroll', this.handleScroll, false);

  it('shallow renders without crashing', () => {
    const { currentPage, error, fetchItems, lastPage, requested, items } = getDefaultProps();
    shallow(
      <InfiniteScrollView
        currentPage={currentPage}
        error={error}
        fetchItems={fetchItems}
        items={items}
        lastPage={lastPage}
        requested={requested}
      />
    );
  });

  it('renders differently with not requested and error', () => {
    const { currentPage, fetchItems, lastPage, items, requested } = getDefaultProps();
    const wrapper = shallow(
      <InfiniteScrollView
        currentPage={currentPage}
        error={true}
        fetchItems={fetchItems}
        items={items}
        lastPage={lastPage}
        requested={requested}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls fetchTodos on mount', () => {
    const { currentPage, error, fetchItems, items, lastPage, requested } = getDefaultProps();
    shallow(
      <InfiniteScrollView
        currentPage={currentPage}
        error={error}
        fetchItems={fetchItems}
        items={items}
        lastPage={lastPage}
        requested={requested}
      />
    );
    expect(fetchItems.mock.calls.length).toBe(1);
    expect(fetchItems.mock.calls[0][0]).toBe(currentPage);
  });

  /*
  it('calls fetchItems on next click', () => {
    const { currentPage, error, fetchItems, items, requested } = getDefaultProps();
    const wrapper = shallow(
      <PaginatedView
        currentPage={currentPage}
        error={error}
        fetchItems={fetchItems}
        items={items}
        lastPage={1}
        requested={requested}
      />
    );
    const testNext = wrapper.find(`#${styles.next}`);
    testNext.simulate('click');
    const callsLength = 2;
    const nextPage = 1;
    expect(fetchItems.mock.calls).toHaveLength(callsLength);
    expect(fetchItems.mock.calls[1][0]).toBe(nextPage);
  });

  it('calls fetchItems on previous click', () => {
    const { error, fetchItems, items, requested } = getDefaultProps();
    const wrapper = shallow(
      <PaginatedView
        currentPage={1}
        error={error}
        fetchItems={fetchItems}
        items={items}
        lastPage={1}
        requested={requested}
      />
    );
    const testPrevious = wrapper.find(`#${styles.previous}`);
    testPrevious.simulate('click');
    const callsLength = 2;
    const previousPage = 0;
    expect(fetchItems.mock.calls).toHaveLength(callsLength);
    expect(fetchItems.mock.calls[1][0]).toBe(previousPage);
  });
  */
});
