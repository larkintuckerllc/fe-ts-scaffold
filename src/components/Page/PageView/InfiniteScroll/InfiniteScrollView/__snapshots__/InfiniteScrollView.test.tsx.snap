// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScrollView component renders differently with not requested and error 1`] = `
<div>
  Error
</div>
`;
